File: redsheet-portal/src/app/database-models/project.ts
import { required, maxLength, range, nested } from '@rx/annotations';
import { TemplateGroup, RecentActivityAndNotification, ProjectModule, vProjectRecord } from './'
import { ProjectModuleAssigneesOrReviewer } from './project-module-assignees-or-reviewer';

export class Project {
	constructor(project?: Project | vProjectRecord) {
		let properties = ["createdBy", "createdOn", "isAllowCustomization", "isClosed", "isStarted", "moduleIdReached", "noOfDays", "ownerId", "projectId", "projectName", "projectNote", "publishDate", "reporteeName", "status", "templateId", "templateName"
			, "updatedBy", "updatedOn", "negotiationRoleId", "templateGroupId", "recentActivityAndNotifications", "projectModules", "projectModuleAssigneesOrReviewers"];

		for (let property of properties)
			if (project && project[property])
				this[property] = project[property];
	}

	createdBy: number = undefined;

	createdOn: Date = undefined;

	isAllowCustomization: boolean = false;

	isClosed: boolean = false;

	isStarted: boolean = false;

	moduleIdReached: number = undefined;

	@range(1, 2147483647)
	noOfDays: number = undefined;

	@range(1, 2147483647)
	ownerId: number = undefined;

	projectId: number = 0;

	@required()
	@maxLength(200)
	projectName: string = undefined;

	@maxLength(1000)
	projectNote: string = undefined;

	publishDate: Date = undefined;

	@maxLength(200)
	reporteeName: string = undefined;

	status: boolean = false;

	@range(1, 2147483647)
	templateId: number = undefined;

	@required()
	@maxLength(200)
	templateName: string = undefined;

	updatedBy: number = undefined;

	updatedOn: Date = undefined;

	@range(0, 2147483647)
	negotiationRoleId: number = undefined;

	@required()
	templateGroupId: string = undefined;
	templateGroup: TemplateGroup;
	@nested(RecentActivityAndNotification)
	recentActivityAndNotifications: RecentActivityAndNotification[];

	@nested(ProjectModule)
	projectModules: ProjectModule[];

	@nested(ProjectModuleAssigneesOrReviewer)
	projectModuleAssigneesOrReviewers: ProjectModuleAssigneesOrReviewer[];

}

File: RedSheet.DbEntities/Models/Project.cs

using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using Microsoft.EntityFrameworkCore;
using Rx.Core.Data.Attributes;
using RedSheet.DbEntities.Enums;

namespace RedSheet.DbEntities.Models
{
    [Table("Projects", Schema = "dbo")]
    [Serializable]
    public partial class Project
    {

        public Nullable<int> CreatedBy { get; set; }

        public Nullable<System.DateTime> CreatedOn { get; set; }

        [Required]
        public bool IsAllowCustomization { get; set; }

        [Required]
        public bool IsClosed { get; set; }

        [Required]
        public bool IsStarted { get; set; }

        public Nullable<int> ModuleIdReached { get; set; }

        [Range(1, int.MaxValue)]
        public int NoOfDays { get; set; }

        [Range(1, int.MaxValue)]
        public int OwnerId { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Key]
        public int ProjectId { get; set; }

        [Required]
        [MaxLength(200)]
        public string ProjectName { get; set; }

        [MaxLength(1000)]
        public string ProjectNote { get; set; }

        public Nullable<System.DateTime> PublishDate { get; set; }

        [MaxLength(200)]
        public string ReporteeName { get; set; }

        [Required]
        public bool Status { get; set; }

        [Range(1, int.MaxValue)]
        public int TemplateId { get; set; }

        [Required]
        [MaxLength(200)]
        public string TemplateName { get; set; }

        public Nullable<int> UpdatedBy { get; set; }

        public Nullable<System.DateTime> UpdatedOn { get; set; }

        [RelationshipTableAttribue("ApplicationObjects", "dbo")]
        public NegotiationRole NegotiationRoleId { get; set; }

        [Required]
        [RelationshipTableAttribue("TemplateGroups", "dbo")]
        //Relationships
        public Guid TemplateGroupId { get; set; }

        [ForeignKey("TemplateGroupId")]
        public TemplateGroup TemplateGroup { get; set; }
        [InverseProperty("Project")]
        public ICollection<ProjectModule> ProjectModules { get; set; }

        [InverseProperty("Project")]
        public ICollection<ProjectModuleAssigneesOrReviewer> ProjectModuleAssigneesOrReviewers { get; set; }

        [InverseProperty("Project")]
        public ICollection<RecentActivityAndNotification> RecentActivityAndNotifications { get; set; }

        public Project()
        {
            this.ProjectModules = new HashSet<ProjectModule>();

            this.RecentActivityAndNotifications = new HashSet<RecentActivityAndNotification>();

            this.ProjectModuleAssigneesOrReviewers = new HashSet<ProjectModuleAssigneesOrReviewer>();
        }
    }
}

